pulsar-common/src/main/java/org/apache/pulsar/common/util/RateLimiter.java

```diff
@@ -0,0 +1,175 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.util;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * 
+ * A Rate Limiter that distributes permits at a configurable rate. Each {@link #acquire()} blocks if necessary until a
+ * permit is available, and then takes it. Each {@link #tryAcquire()} tries to acquire permits from available permits,
+ * it returns true if it succeed else returns false. Rate limiter release configured permits at every configured rate
+ * time, so, on next ticket new fresh permits will be available.
+ * <p>
+ * For example: if RateLimiter is configured to release 10 permits at every 1 second then RateLimiter will allow to
+ * acquire 10 permits at any time with in that 1 second.
+ *
+ */
+public class RateLimiter {
```

Comment:
Ok, I found out the gist with the explanation and it makes sense. Can you include the full explanation here as well? Also, would it make sense to have the `1sec` period to be configurable as well? Eg: only rate limit if it's exceeding the rate for 10 seconds instead of 1.